network/hybridNetwork_test.go

```diff
@@ -0,0 +1,148 @@
+
+	relayCfg := cfg
+	relayCfg.ForceRelayMessages = true
+	netA, err := NewHybridP2PNetwork(log.With("node", "netA"), relayCfg, p2pKeyDir, nil, genesisID, "net", &nopeNodeInfo{})
+	require.NoError(t, err)
+
+	err = netA.Start()
+	require.NoError(t, err)
+
+	// collect ws address
+	addr, portListen := netA.wsNetwork.Address()
+	require.True(t, portListen)
+	require.NotZero(t, addr)
+	parsed, err := url.Parse(addr)
+	require.NoError(t, err)
+	addr = parsed.Host
+	netA.Stop()
+
+	// make it net address and restart the node
+	relayCfg.NetAddress = addr
+	relayCfg.PublicAddress = addr
+	netA, err = NewHybridP2PNetwork(log.With("node", "netA"), relayCfg, p2pKeyDir, nil, genesisID, "net", &nopeNodeInfo{})
+	require.NoError(t, err)
+
+	err = netA.Start()
+	require.NoError(t, err)
+	defer netA.Stop()
+
+	// collect relay address and prepare nodes phonebook
+	peerInfoA := netA.p2pNetwork.service.AddrInfo()
+	addrsAp2p, err := peer.AddrInfoToP2pAddrs(&peerInfoA)
+	require.NoError(t, err)
+	require.NotZero(t, addrsAp2p[0])
+	multiAddrStr := addrsAp2p[0].String()
+
+	fullAddr, portListen := netA.wsNetwork.Address()
+	require.True(t, portListen)
+	require.NotZero(t, addr)
+	require.Contains(t, fullAddr, addr)
+
+	phoneBookAddresses := []string{multiAddrStr, addr}
+
+	netB, err := NewHybridP2PNetwork(log.With("node", "netB"), cfg, "", phoneBookAddresses, genesisID, "net", &nopeNodeInfo{})
+	require.NoError(t, err)
+	// for netB start the p2p network first
+	err = netB.p2pNetwork.Start()
+	require.NoError(t, err)
+	defer netB.Stop()
+
+	netC, err := NewHybridP2PNetwork(log.With("node", "netC"), cfg, "", phoneBookAddresses, genesisID, "net", &nopeNodeInfo{})
+	require.NoError(t, err)
+	// for netC start the ws network first
+	err = netC.wsNetwork.Start()
+	require.NoError(t, err)
+	defer netC.Stop()
+
+	// ensure initial connections are done
+	require.Eventually(t, func() bool {
+		return len(netA.GetPeers(PeersConnectedIn)) == 2
+	}, 3*time.Second, 50*time.Millisecond)
+
+	// start the second half of the hybrid net
+	err = netB.wsNetwork.Start()
+	require.NoError(t, err)
+	err = netC.p2pNetwork.Start()
+	require.NoError(t, err)
+
+	// wait for connection attempts. nodes need some time to make connections,
+	// and instead of `time.Sleep(1 * time.Second)` the networkPeerIdentityDisconnect net identity counter is used.
+	// Since this test is not parallel the networkPeerIdentityDisconnect should not be modified from outside.
+	// Both netB and netC are attempting to connect but netA could also open an outgoing stream in netB or netC connection.
+	// So, the counter should be at least 2+identDiscValue.
+	require.Eventually(t, func() bool {
+		return networkPeerIdentityDisconnect.GetUint64Value() >= 2+identDiscValue
+	}, 3*time.Second, 50*time.Millisecond)
+
+	// now count connections
+	// netA should have 2 connections, not 4
+	// netB should have 1 connection (via p2p)
+	// netC should have 1 connection (via ws)
```

Comment:
General question: this PR cancels the newer connection to/from a peer if a duplicate is made. I wonder if it would be beneifitial to prioritize the p2p connection and drop the ws one in this case, since both nodes are p2p capable. Though I realize that would be more complicated.

Reply:
Actually, it is already a case:
1. Hybrid net starts p2p net first so it implicitly takes priority. Yes, it is racy but it is what happen most of the time.
2. dedup work only on websocket network so p2p conns are prioritized.

Reply:
Sure this scenario is possible when libp2p drops connection to A and wsnet connects to in by doing routine conns rotation. Both p2p and ws networks have connection targets so as soon as there are enough relays both nets will maintain some outgoing conns.

Reply:
And the dedup work going on elsewhere will address any lingering concerns on that front